[UI-side compositing] Add logging for painting new tile for ScrollPerf

nmoucht opened the pull request:
#### d4f1d3db3f612fc1283129792b44882a657f7fa5
<pre>
[UI-side compositing] Add logging for painting new tile for ScrollPerf
<a href="https://bugs.webkit.org/show_bug.cgi?id=254050">https://bugs.webkit.org/show_bug.cgi?id=254050</a>
rdar://106775893

Reviewed by Simon Fraser.

Use iOS&apos;s RemoteLayerTreeScrollingPerformanceData to calculate blank tiles
for new tile creation and scrolling. Log all of the accumulated log events
when the WebPageProxy is closed. Also, disable the logging on TileController
to prevent double logging new tile events in the web process by having
PlatformCALayerRemoteTiledBacking set a bool.

* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameScrollPosition const):
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::TileController):
(WebCore::TileController::logFilledVisibleFreshTile):
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::platformCALayerPaintContents):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::gatherPlatformLayerList):
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::lastViewportRect const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::willCommitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::setViewportRect):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::hadFirstPaint):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::reportFilledVisibleFreshTile):
(WebKit::RemoteScrollingCoordinatorProxy::setScrollingPerformanceTestingEnabled const):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingPerformanceTestingEnabled const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:

Canonical link: <a href="https://commits.webkit.org/262258@main">https://commits.webkit.org/262258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c02ceec9c77a5c0630c73a32b0e9c063d1be8c0d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1059 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1098 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1601 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/909 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1108 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1139 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1107 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1037 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/961 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1505 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1007 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/959 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/936 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/987 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/982 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/923 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/968 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/998 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/106 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/ca/TileController.h:
```diff
@@ -55,7 +55,7 @@ class TileController final : public TiledBacking {
     friend class TileGrid;
 public:
-    WEBCORE_EXPORT explicit TileController(PlatformCALayer*);
+    WEBCORE_EXPORT explicit TileController(PlatformCALayer*, bool = true);
```
The argument needs a name.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/ca/TileController.cpp:
```diff
@@ -794,7 +795,8 @@ void TileController::removeUnparentedTilesNow()
 void TileController::logFilledVisibleFreshTile(unsigned blankPixelCount)
 {
-    owningGraphicsLayer()->platformCALayerLogFilledVisibleFreshTile(blankPixelCount);
+    if (!m_shouldLogScrollingPerformance)
```
I am confused by the `!`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/ca/TileGrid.cpp:
```diff
@@ -720,7 +720,7 @@ void TileGrid::platformCALayerPaintContents(PlatformCALayer* platformCALayer, Gr
 
         if (repaintCount == 1 && !visiblePart.isEmpty())
-            m_controller.logFilledVisibleFreshTile(blankPixelCount());
+            m_controller.logFilledVisibleFreshTile(999999);
```
Doesn't seem right

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -186,6 +190,28 @@ static void applyCommonPropertiesToLayer(CALayer *layer, const RemoteLayerTreeTr
+        auto sublayers = adoptNS([[layer sublayers] copy]);
+
+        // If this layer is the parent of a tile, it is the parent of all of the tiles and nothing else.
+        if ([[[sublayers objectAtIndex:0] valueForKey:@"isTile"] boolValue]) {
```
We should probably just set a property on the tile grid container layers so this logic can be simpler.

nmoucht replied:
I think we want to use all of the related layers for the `RemoteLayerTreeHost` (we want what is analogous to `m_tiles` in `TileGrid`)

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -130,6 +131,12 @@
 void RemoteLayerTreeDrawingAreaProxy::willCommitLayerTree(TransactionID transactionID)
 {
     m_pendingLayerTreeTransactionID = transactionID;
+    
```
Extra line

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -130,6 +131,12 @@
+
+void RemoteLayerTreeDrawingAreaProxy::setViewportRect(const RemoteLayerTreeTransaction& transaction)
+{
+    m_lastViewportRect = LayoutRect(transaction.scrollPosition(),  transaction.baseLayoutViewportSize());
```
Whitespace issues on this line

nmoucht replied:
Removed this.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -130,6 +131,12 @@
+    
+}
+
+void RemoteLayerTreeDrawingAreaProxy::setViewportRect(const RemoteLayerTreeTransaction& transaction)
```
The name is a bit odd since the argument isn't the viewport rect. Maybe "store viewport rect"?

nmoucht replied:
Removed this function.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -166,7 +173,8 @@
 #endif
     };
-
+    m_webPageProxy.scrollingCoordinatorProxy()->setScrollingPerformanceTestingEnabled(const_cast<RemoteScrollingCoordinatorTransaction&>(scrollingTreeTransaction).scrollingStateTree()->rootStateNode()->scrollingPerformanceTestingEnabled());
```
It isn't right to to reach into the state tree. The ScrollingTree gets it (see `ScrollingTree::commitTreeState`) so grab it from there the way we do for other tree-wide state.

nmoucht replied:
This was a workaround due to the original solution logging directly in `applyCommonPropertiesToLayer` where on the first commit we wouldn't have scrolling tree state. Now we tally up layers that had layer backing stores applied to them and log all of them once the commit is finished.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
```diff
@@ -98,6 +98,7 @@ class RemoteLayerTreeNode : public CanMakeWeakPtr<RemoteLayerTreeNode> {
     {
         m_cachedContentsBuffers = WTFMove(buffers);
     }
+    bool hadFirstPaint() { return std::exchange(m_hadFirstPaint, true); }
```
Blank line below. Consider a name that sounds less like a pure getter.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
```diff
@@ -380,6 +385,18 @@ void RemoteScrollingCoordinatorProxy::removeWheelEventTestCompletionDeferralForR
+{
+    m_scrollingTree->setScrollingPerformanceTestingEnabled(flag);
+}
+
```
Extra blank line here and below.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/ca/TileController.cpp:
```diff
@@ -63,13 +63,14 @@ String TileController::zoomedOutTileGridContainerLayerName()
 }
 
-TileController::TileController(PlatformCALayer* rootPlatformLayer)
+TileController::TileController(PlatformCALayer* rootPlatformLayer, bool shouldLogScrollingPerformance)
```
This is "allowScrollPerformanceLogging". Passing `true` doesn't turn on logging, unless the pref is also set.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/ca/TileController.h:
```diff
@@ -241,6 +241,8 @@ class TileController final : public TiledBacking {
     bool m_haveExternalVelocityData { false };
     bool m_isTileSizeUpdateDelayDisabledForTesting { false };
+    
+    bool m_shouldLogScrollingPerformance { false };
```
`allow`

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteTiledBacking.cpp:
```diff
@@ -36,7 +36,7 @@ using namespace WebCore;
 PlatformCALayerRemoteTiledBacking::PlatformCALayerRemoteTiledBacking(LayerType layerType, PlatformCALayerClient* owner, RemoteLayerTreeContext& context)
     : PlatformCALayerRemote(layerType, owner, context)
-    , m_tileController(makeUnique<TileController>(this))
+    , m_tileController(makeUnique<TileController>(this, false))
```
Boolean trap here (hard to know what `false` means). Would be good to either add a comment, or use an enum.

nmoucht replied:
Changed to use enum.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeScrollingPerformanceData.h:
```diff
@@ -40,34 +41,39 @@ class RemoteLayerTreeScrollingPerformanceData {
 
     void didCommitLayerTree(const WebCore::FloatRect& visibleRect);
     void didScroll(const WebCore::FloatRect& visibleRect);
+    void didChangeSynchronousScrollingReasons(WTF::MonotonicTime, uint64_t);
```
The `uint64_t` is very mysterious. What is it?

nmoucht replied:
Added name to clarify.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeScrollingPerformanceData.h:
```diff
@@ -40,34 +41,39 @@ class RemoteLayerTreeScrollingPerformanceData {
 
     NSArray *data(); // Array of [ time, event type, unfilled pixel count ]
-
+    void logData();
```
Add blank line below.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeScrollingPerformanceData.h:
```diff
@@ -40,34 +41,39 @@ class RemoteLayerTreeScrollingPerformanceData {
+        WTF::MonotonicTime endTime;
         EventType eventType;
-        unsigned blankPixelCount;
+        uint64_t logData;        
```
Maybe `value` ?

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeScrollingPerformanceData.h:
```diff
@@ -40,34 +41,39 @@ class RemoteLayerTreeScrollingPerformanceData {
         { }
         
-        bool canCoalesce(BlankPixelCount::EventType, unsigned blankPixelCount) const;
+        bool canCoalesce(ScrollingLogEvent::EventType, unsigned blankPixelCount) const;
```
`unsigned` should be `uint64_t`

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeScrollingPerformanceData.h:
```diff
@@ -40,34 +41,39 @@ class RemoteLayerTreeScrollingPerformanceData {
-    void appendBlankPixelCount(BlankPixelCount::EventType, unsigned blankPixelCount);
+    unsigned blankPixelCount(const WebCore::FloatRect& visibleRect) const;
+
+    void appendBlankPixelCount(ScrollingLogEvent::EventType, unsigned blankPixelCount);
```
Might as well use `uint64_t` everywhere.

nmoucht replied:
Fixed.

nmoucht committed:
[UI-side compositing] Add logging for painting new tile for ScrollPerf
https://bugs.webkit.org/show_bug.cgi?id=254050
rdar://106775893

Reviewed by Simon Fraser.

Use iOS's RemoteLayerTreeScrollingPerformanceData to calculate blank tiles
for new tile creation and scrolling. Log all of the accumulated log events
when the WebPageProxy is closed. Also, disable the logging on TileController
to prevent double logging new tile events in the web process by having
PlatformCALayerRemoteTiledBacking set a bool.

* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameScrollPosition const):
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::TileController):
(WebCore::TileController::logFilledVisibleFreshTile):
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::platformCALayerPaintContents):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::gatherPlatformLayerList):
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::lastViewportRect const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::willCommitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::setViewportRect):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::hadFirstPaint):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::reportFilledVisibleFreshTile):
(WebKit::RemoteScrollingCoordinatorProxy::setScrollingPerformanceTestingEnabled const):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingPerformanceTestingEnabled const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:

Canonical link: https://commits.webkit.org/262258@main
